### PR TITLE
make local-up-hypernetes.sh robuster

### PR DESCRIPTION
**Pull request by Crazykev**
1. mainly make local-up-hypernetes.sh robuster(theoretically, rerun script when encounter failure will not fail)
2. fixed some bugs
3. add DEV_MODE for development environment

cc @feiskyer 

Signed-off-by: HaoZhang crazykev@zju.edu.cn

### CODE REVIEW

File: hack/hypernetes/ceph.sh [hunk 1]
```diff
@@ -20,10 +20,24 @@ set -o pipefail
 
 function kube::util::setup_ceph() {
 	echo "Start $FUNCNAME"
+	rm -rf /etc/ceph
+	mkdir -p /etc/ceph
+	if ! kube::util::do_setup_ceph ; then
+		echo "Preinstall some package before exectue ceph-deploy..."
+		kube::util::ceph_preinstall
+		kube::util::do_setup_ceph
```
Comment (feiskyer) [inline hack/hypernetes/ceph.sh:28]: I think just calling one `kube::util::do_setup_ceph` is enough here. If the install is failed, the script could be run again.

File: hack/hypernetes/hypernetes.sh [hunk 2]
```diff
@@ -28,21 +28,22 @@ function kube::util::download_hypernetes() {
 
 function kube::util::build_hypernetes() {
 	echo "Start $FUNCNAME"
-	cd ${GOPATH}/src/k8s.io/kubernetes
+	cd ${GO_K8S_ROOT}/hypernetes
```
Comment (feiskyer) [inline hack/hypernetes/hypernetes.sh]: The repo dir should be `${GOPATH}/src/k8s.io/kubernetes`

File: hack/hypernetes/hyper.sh [hunk 3]
```diff
@@ -29,7 +29,9 @@ group = "root"
 clear_emulator_capabilities = 0
 EOF
 
-	curl -sSL http://hypercontainer.io/install | bash
+	if curl -sSl http://hypercontainer.io/install | bash || curl -sSL http://hypercontainer.io/install | bash | grep "You are using the newest hyper" ; then
```
Comment (feiskyer) [inline hack/hypernetes/hyper.sh]: Could we extract the check out here? e.g. check whether hyperctl is installed locally and check its version. If not latest, then install via `http://hypercontainer.io/install`.

File: hack/hypernetes/hypernetes.sh [hunk 4]
```diff
@@ -28,21 +28,22 @@ function kube::util::download_hypernetes() {
...
 
+	rm -rf /var/lib/kubernetes /srv/kubernetes /var/log/kubernetes /var/run/kubernetes
 	mkdir -p /var/lib/kubernetes /srv/kubernetes  /var/log/kubernetes /var/run/kubernetes
 	chown kube:kube /var/log/kubernetes/
 	chown kube:kube /var/run/kubernetes/
 	chown kube:kube /srv/kubernetes
 	# openssl genrsa -out /var/lib/kubernetes/serviceaccount.key 2048
 	# chown kube:kube /var/lib/kubernetes/serviceaccount.key
-	CERT_GROUP=kube CERT_DIR=/srv/kubernetes $GOPATH/src/k8s.io/kubernetes/cluster/saltbase/salt/generate-cert/make-ca-cert.sh $(hostname -i) IP:10.254.0.1,IP:${IF_IP},DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:k8s.local
+	CERT_GROUP=kube CERT_DIR=/srv/kubernetes ${GO_K8S_ROOT}/hypernetes/cluster/saltbase/salt/generate-cert/make-ca-cert.sh $(hostname -i) IP:10.254.0.1,IP:${IF_IP},DNS:kubernetes,DNS:kubernetes.default,DNS:kubernetes.default.svc,DNS:kubernetes.default.svc.cluster.local,DNS:k8s.local
```
Comment (feiskyer) [inline hack/hypernetes/hypernetes.sh]: Again, should be `k8s.io/kubernetes`

File: hack/hypernetes/openstack.sh [hunk 5]
```diff
@@ -47,9 +49,23 @@ function kube::util::setup_openstack() {
...
 	sed -i 's/#extension_drivers.*/extension_drivers = port_security,dns/g' /etc/neutron/plugins/ml2/ml2_conf.ini
 	systemctl restart neutron-server
 }
+
+function kube::util::clean_network() {
+	local net_name=${1:-br-ex}
+	net_list=`neutron net-list -f value -c id --name ${net_name}`
+	if [[ -n ${net_list-} ]] ; then
+		for n in net_list ; do
+			neutron net-delete n
```
Comment (feiskyer) [inline hack/hypernetes/openstack.sh]: should be `${net_list}`, `$n`

File: hack/hypernetes/util.sh [hunk 6]
```diff
@@ -36,8 +36,12 @@ function kube::util::verify_system() {
 
 function kube::util::setup_ssh() {
 	echo "Start $FUNCNAME"
-	ssh-keygen -f /root/.ssh/id_rsa -t rsa -N ''
-	cat /root/.ssh/id_rsa.pub >> /root/.ssh/authorized_keys
+	local key_path="/root/.ssh/id_rsa"
+        if [[ -e ${key_path} ]] ; then
+		rm -f ${key_path}
```
Comment (feiskyer) [inline hack/hypernetes/util.sh]: This is a very dangerous operation, we shouldn't remove the ssh key. Instead, if the key already exist, just jump to `authorized_keys` and `ssh-keyscan` step.

File: hack/hypernetes/util.sh [hunk 7]
```diff
@@ -104,3 +108,27 @@ function kube::util::get_ip() {
...
+		echo "Git repo ${repo_path} already exist, leave it unchange in dev mode"
+	else
+		rm -rf ${repo_path}
+		git clone ${repo_url} ${repo_path}
+	fi
+}
+
+function kube::util::ensure_dir_exist() {
+	local dir=$1
+	if ! [[ -d ${dir} ]] ; then
```
Comment (feiskyer) [inline hack/hypernetes/util.sh]: This check is actually unnecessary, `mkdir -p` will always returns success regardless of whether the dir is already exist.

**Comment by resouer** [comment]
Another nit, if you want to execute the script multiple times, you need to check whether selinux is already disabled, otherwise `setenforce 0` would fail.

Ref: https://www.centos.org/docs/5/html/5.2/Deployment_Guide/sec-sel-enable-disable-enforcement.html

File: hack/hypernetes/ceph.sh (hunk 1 above)
Comment (Crazykev) [reply hack/hypernetes/ceph.sh:28]: @feiskyer It's just a workaround here. Actually these two approach will install `ceph` from different repo, `kube::util::ceph_preinstall` will install from centos repo, ie: mirrors from aliyun, while `do_setup_ceph` will install `ceph` from official repo through `ceph-deploy`.  The second one is very slow and may always fail in somewhere like China.

**Comment by Crazykev** [comment]
@feiskyer @resouer Updated. 

> you need to check whether selinux is already disabled, otherwise setenforce 0 would fail.

I didn't see this problem in local environment. For safety, I'll do a check before setting the value.

**Comment by feiskyer** [comment]
lgtm. @resouer do you have more comments?

**Comment by resouer** [comment]
lgtm too, nice work

**Pull request merged**